feat: bank reconciliation rule

0xD0M1M0 opened the pull request:
This PR adds two main functionallities and a side functionallity:

1. Book included bank fees to a selected account
2. Book bank transactions based on preset rules
3. Convert all negative values for deposit, withdrawal and included_fee/excluded fee into positive values

The order of applied functions is 3, 1 and 2. All values are converted into positive values, then bank fees are booked and if unreconciled values remain, bank reconciliation rules are applied.

**Function 1:**
Bank fees saved in the included_fee field in ERPNext (depends https://github.com/frappe/erpnext/pull/51112) are booked on submit to a specific bank fee account. This account is specified in the _"Bank Account"_.

**Function 2:**
On submit, the bank transaction is checked if it matches one pre-set booking rule in the DocType _"Bank Reconciliation Rule"_. If it matches, the full unallocated amount is booked to an _"Account"_ specified in the rule. Partical bookings are not supported. Also no matching with invoices is done. It purely books the bank transaction to an Account.

**Function 3:**
This support function eliminates import errors as all values in the fields _deposit_, _withdrawal_, _included_fee_ and _excluded fee_ need to be positive.

This PR works also with Bank Statement imports and bank imports, as they are applied to the Bank Transaction directly.

To Do:

- [ ] Mention `enforce_positive_values` in documentation (after merge)
- [x] Put both Fee and Reco Rule into the same Journal Entry(?) -> No
- [x] Link Fee Journal Entry to Bank Transaction, so they are traceable and cancelled together

0xD0M1M0 committed:
feat: bank reconciliation rule

0xD0M1M0 committed:
fix: recall handle_excluded_fee

PatrickDEissler commented on banking/klarna_kosma_integration/doctype/bank_reconciliation_rule/bank_reconciliation_rule.py:
```diff
@@ -0,0 +1,18 @@
+		bank_account_currency = frappe.db.get_value("Account", bank_account, "account_currency")
+		target_account_currency = frappe.db.get_value("Account", self.target_account, "account_currency")
+		if bank_account_currency != target_account_currency:
+			frappe.throw(_("Bank Account and Target Account need to be in the same currency!"))
```
refactor: I suggest to wrap this into a separate function.

PatrickDEissler commented on banking/overrides/bank_account.py:
```diff
@@ -1,4 +1,16 @@
+		bank_account_currency = frappe.db.get_value("Account", doc.account, "account_currency")
+		bank_fee_currency = frappe.db.get_value("Account", doc.bank_fee_account, "account_currency")
+		if bank_account_currency != bank_fee_currency:
+			frappe.throw(_("Company Account and Bank Fee Account must be in the same currency!"))
```
refactor: I suggest to wrap this into a separate function, as well.

PatrickDEissler commented on banking/overrides/bank_transaction.py:
```diff
@@ -72,3 +76,251 @@ def on_update_after_submit(doc, event):
+	ensure_positive_deposit_withdrawal_fees(doc, method)
+
+
+def on_cancel(doc, method):
```
refactor: it is easier if the run-methods occur in the py-file in sequential order. Meaning: The `on_cancel` method should occur after `before_submit`.

PatrickDEissler commented on banking/overrides/bank_transaction.py:
```diff
@@ -72,3 +76,251 @@ def on_update_after_submit(doc, event):
+			if doc.docstatus == 1:
+				doc.cancel()
+		except Exception as e:
+			frappe.msgprint(f"Failed to cancel {journal_entry}: {e}")
```
```suggestion
			frappe.msgprint(_("Failed to cancel {0}: {1}").format(journal_entry, e))
```
That is how it is translatable. Note: The syntax of the suggestion might be a bit off. Please, check it again.

PatrickDEissler commented on banking/overrides/bank_transaction.py:
```diff
@@ -72,3 +76,251 @@ def on_update_after_submit(doc, event):
+			frappe.msgprint(f"Failed to cancel {journal_entry}: {e}")
+
+
+def before_submit(doc, method):
```
refactor: the run method runs a lot of code directly. Consider using separate functions to organize the code.

PatrickDEissler commented on banking/overrides/bank_transaction.py:
```diff
@@ -72,3 +76,251 @@ def on_update_after_submit(doc, event):
+		)
+
+
+def create_automatic_journal_entry(doc, company_doc, date, account, target_account, rule, debit, credit):
```
A lot of things are repeated in create_automatic_journal_entry and create_fee_journal_entry. 
Consider using only one method for both use cases.

PatrickDEissler commented on banking/overrides/bank_transaction.py:
```diff
@@ -72,3 +76,251 @@ def on_update_after_submit(doc, event):
+def create_automatic_journal_entry(doc, company_doc, date, account, target_account, rule, debit, credit):
+	journal_entry = frappe.new_doc("Journal Entry")
+	journal_entry.voucher_type = "Journal Entry"
+	journal_entry.title = f"BT ID {doc.name}"
```
I don't think this PR should affect the title.

0xD0M1M0 committed:
refactor: refactor functions

0xD0M1M0 committed:
chore: remove comments

0xD0M1M0 committed:
fix: create_je_bank_fees only if included fee is set

0xD0M1M0 committed:
fix: bank_account

0xD0M1M0 committed:
fix: deposit fees and rule credit value corrected

0xD0M1M0 committed:
feat: added test cases

barredterra committed:
fix(Bank Account): various fixes

- filter by broader "Expense" type instead of "COGS"
- Show Fee account only when Company Account is set (se we can always fetch the relevant filter values)
- Fetch company and currency in the same query

barredterra committed:
feat: add rule to workspace

barredterra committed:
fix(Bank Reconciliation Rule): filter_description, type hints

barredterra committed:
fix(Bank Reconciliation Rule): show only bank accounts with company account

barredterra committed:
refactor(Bank Reconciliation Rule): set target account query

barredterra committed:
refactor(Bank Reconciliation Rule): render_bt_filters

barredterra committed:
feat(Bank Reconciliation Rule): make filters read-only when submitted

0xD0M1M0 committed:
fix: set clearance date after submit

clearance_date is set to None during submit, set afterwards to clear journal entry

barredterra committed:
refactor: test case for Bank Reconciliation Rule

barredterra commented on banking/overrides/bank_transaction.py:
```diff
@@ -72,3 +108,154 @@ def on_update_after_submit(doc, event):
+	doc.deposit = abs(flt(doc.deposit) or 0.0)
+	doc.withdrawal = abs(flt(doc.withdrawal) or 0.0)
+	doc.included_fee = abs(flt(doc.included_fee) or 0.0)
+	doc.excluded_fee = abs(flt(doc.excluded_fee) or 0.0)
```
The only falsy value that can come out of `flt` is `0.0`, so saying "zero or zero" seems redundant.

> Convert all negative values for deposit, withdrawal and included_fee/excluded fee into positive values

Also, if you don't mind, I'd like to exclude this feature from this PR. It doesn't seem to be a prerequisite for the other features and the design seems quite opinionated. (You could argue that a negative deposit is equivalent to a withdrawal, not to a positive deposit.)

0xD0M1M0 replied:
> Also, if you don't mind, I'd like to exclude this feature from this PR. It doesn't seem to be a prerequisite for the other features and the design seems quite opinionated. (You could argue that a negative deposit is equivalent to a withdrawal, not to a positive deposit.)

Yes, that argument is valid. In practice, negative values in any of these fields often lead to incorrect data handling in subsequent steps, which is usually unintended. Bank statements come in many different formats. In my experience, some daily business bank statements include both columns for deposits and withdrawals, but the withdrawal values are still marked as negative. During import via the bank statement import, this can directly lead to incorrect data (see https://github.com/alyf-de/banking/issues/189).

My suggestion: keep the current function as is, and add a second function in a separate field to handle bank account data from statements that use a single column format, where negative values indicate withdrawals and positive values indicate deposits. This would cover a second use case. For example, some German bank CSV exports use entries like "-100,00 EUR". With this approach, the content can be correctly parsed and assigned to the appropriate fields using absolute values.

barredterra committed:
refactor: move exceptions and helpers to general utils

barredterra committed:
refactor: test_bank_account to make use of helpers

barredterra committed:
refactor: test_bank_transaction to make use of helpers

barredterra committed:
refactor: don't check for exact error message

barredterra committed:
refactor: move doc creation into helpers

barredterra commented on banking/custom_fields.py:
```diff
@@ -87,4 +87,14 @@ def get_custom_fields():
+			dict(
+				fieldname="bank_fee_account",
+				fieldtype="Link",
+				label="Bank Fee Account",
```
```suggestion
				label=_("Bank Fee Account"),
```

Labels need to be marked as translatable

barredterra committed:
Merge remote-tracking branch 'upstream/version-15-hotfix' into bank-reconcilation-rule

barredterra committed:
fix: user-facing messages

barredterra committed:
chore: update translations

barredterra committed:
fix: translate field labels

barredterra committed:
fix: mark label as translatable

barredterra committed:
refactor(Bank Transaction): enforce positive values

barredterra committed:
refactor: pass cost center instead of company doc

barredterra committed:
fix: handle case when fee is not set

barredterra committed:
test: don't convert None

barredterra committed:
test: make sure we use the same company everywhere

barredterra committed:
refactor: create_je_automatic_rules

- reduce nesting
- catch JSONDecodeErrors

barredterra committed:
fix: avoid comparing with None

barredterra committed:
test: separate class for positive/negative unit test

barredterra committed:
test: centralize setup, rely on rollback for deletion

barredterra committed:
test: centralize setup, rely on rollback for deletion (part 2)

barredterra committed:
Merge branch 'version-15-hotfix' into bank-reconcilation-rule

0xD0M1M0 commented:
In case the bank rule did not work correctly or the accountant decides to assign a different account and cancels the journal entry of the rule, the fee should stay untouched and booked to the fee account. I therefore opt for two separate journal entries. One for the bank fee and one for the rule. In this case the accountant even can use the bank reconciliation tool to reasign the bank transaction without touching the bank fee again.

barredterra committed:
refactor: create_automatic_journal_entry

- type hints
- optional parameters last
- pass atomic values instead of doc

barredterra committed:
fix: make bank fee entry optional

Otherwise existing sites will break because they don't have a Bank Fee Account specified yet.

barredterra commented:
Split into smaller PRs: #345, #346, #347

This way it's easier to reason about and iterate on each feature separately.
